test/unit/core/type_traits/lazy_test.cpp

```diff
@@ -0,0 +1,62 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#include <gtest/gtest.h>
+
+#include <list>
+#include <vector>
+
+#include <seqan3/core/type_traits/lazy.hpp>
+#include <seqan3/std/concepts>
+
+using namespace seqan3;
+
+TEST(lazy, instantiate)
+{
+    EXPECT_TRUE((std::is_same_v<instantiate_t<std::vector<int>>,
+                                std::vector<int>>));
+
+    EXPECT_TRUE((std::is_same_v<instantiate_t<lazy<std::vector, int>>,
+                                std::vector<int>>));
+}
+
+template <typename t>
+    requires std::Integral<t>
+using integral_identity_t = t;
+
+TEST(lazy, lazy_conditional)
+{
+    // regular conditional behaviour
+    EXPECT_TRUE((std::is_same_v<lazy_conditional_t<true,
+                                                   std::true_type,
+                                                   std::false_type>,
+                                std::true_type>));
+    EXPECT_TRUE((std::is_same_v<lazy_conditional_t<false,
+                                                   std::true_type,
+                                                   std::false_type>,
+                                std::false_type>));
+
+    // lazy behaviour, not necessary
+    EXPECT_TRUE((std::is_same_v<lazy_conditional_t<true,
+                                                   lazy<std::vector, int>,
+                                                   lazy<std::list, int>>,
+                                std::vector<int>>));
+    EXPECT_TRUE((std::is_same_v<lazy_conditional_t<false,
+                                                   lazy<std::vector, int>,
+                                                   lazy<std::list, int>>,
+                                std::list<int>>));
+
+    // lazy behaviour, necessary to compile
```

Comment:
I don't think this test case shows the necessity to use `lazy` and `lazy_conditional_t` instead of `std::conditional_t` in order to compile.
```cpp
    EXPECT_TRUE((std::is_same_v<std::conditional_t<true,
                                                   integral_identity_t<int>,
                                                   void>,
                                int>));
```
and
```cpp
    EXPECT_TRUE((std::is_same_v<lazy_conditional_t<true,
                                                   lazy<integral_identity_t, int>,
                                                   void>,
                                int>));
```
Are equivalent. 
BUT
```cpp
    EXPECT_TRUE((std::is_same_v<std::conditional_t<false,
                                                   integral_identity_t<double>,
                                                   int>,
                                int>));
```
will not compile and
```cpp
    EXPECT_TRUE((std::is_same_v<lazy_conditional_t<false,
                                                   lazy<integral_identity_t, double>,
                                                   int>,
                                int>));
```
will compile.